### ISSUE TITLE
UI Slowdown after history information has a few dozen edits to an image

### ISSUE THREAD

**Issue by hafzaal** [darktable-org/darktable#15237]
### Describe the bug

I've noticed that the UI of darktable slows down considerably when an image I'm editing has around 50 items in the history information stack. The UI is very snappy and responsive in the beginning but continues to slow down as I keep on editing a picture for an extended period of time. 

Resetting the history stack to default restores the UI responsiveness. I have tried to reinstall darktable multiple times and have tried versions 4.4.1, 4.4.2 and also the latest nightly build. All of them seem to have this issue. Here is a google drive link that contains an image alongside the .xmp file showing the edits.

https://drive.google.com/file/d/1dF00cEt2VEDKqEWP67trvRnXw4Nf3OU3/view?usp=sharing

### Steps to reproduce

1. Open file in darktable
2. Change any module e.g. exposure or color zones
3. The UI will be slow to respond to changes.

Resetting the history stack of the image should restore darktable's responsiveness.

### Expected behavior

Continue to be responsive (or fairly responsive) even with a larger number of edits.

### Where did you install darktable from?

darktable.org

### darktable version

4.4.2 and nightly build.

### What OS are you using?

Windows

### What is the version of your OS?

Windows 11 Pro 22H2

### Describe your system?

AMD Ryzen 5900x
AMD Radeon 6800XT
Samsung EVO 970 Nvme drive
32GB DDR4 Ram

### Are you using OpenCL GPU in darktable?

Yes

### If yes, what is the GPU card and driver?

6800XT with a variety of drivers. But currently have 23.9.1 installed.

**Comment by apostel338** [comment]
I can confirm your problem. But meanwhile, are you aware of _compress history stack_?

![grafik](https://github.com/darktable-org/darktable/assets/96216629/14dcde93-db64-4569-88f6-c46ab8ec3ced)

This should help and you won't lose your edits as you do when _resetting_.

**Comment by hafzaal** [comment]
> I can confirm your problem. But meanwhile, are you aware of _compress history stack_?
> 
> ![grafik](https://user-images.githubusercontent.com/96216629/267598145-14dcde93-db64-4569-88f6-c46ab8ec3ced.png)
> 
> This should help and you won't lose your edits as you do when _resetting_.

Thanks for that, I didn't ever think to click that button. This does help quite a bit and makes the UI significantly more responsive again.

**Comment by dterrahe** [comment]
This could have something to do with the _undo_ functionality. Every time a change is made (or if there are consecutive changes to the same slider, every 10 secs) a snapshot of the full history stack is saved so it can be rolled back. (It used to be that two copies were made, before and after, for every slider move). This could probably be sped up (by only saving the previous value of the latest change or whatever other action was performed on the history stack, like selecting one of the history points), but that would add complexity and the chance that "everything" gets corrupted, so would have to be looked at by somebody who really understands this stuff.

It could also be something completely different. I think it might be worth it for somebody to run a profiler specifically for user interaction stuff/responsiveness. For the image calculations we have pretty good instrumentation to see where the time is spent, but not so much for interactiveness, in that there's a lot of stuff going on we might not even think to look at.

**Comment by jenshannoschwalm** [comment]
Could also be the preparing of the pipeline - commit_params.

Would be interested in such an xmp file making problems...

**Comment by jenshannoschwalm** [comment]
> Every time a change is made (or if there are consecutive changes to the same slider, every 10 secs) a snapshot of the full history stack is saved so

This would mean to test what is responsiveness like after setting "save history while developing" in preferences to `0`

**Comment by dterrahe** [comment]
Ah sorry, I meant "saved to memory", not disk. Xmp saving is much less frequent so wouldn't be my _first_ suspect. But I'm not excluding anything at all!

commit_params looks at the currently active modules, not how often they've been changed in the history. I read the issue as deteriorating the more often the same modules are fiddled with, leading to a long history, rather than more and more modules being added, because then, yes, I'm not surprised everything slows down ;-).

**Comment by jenshannoschwalm** [comment]
`commit_params` - doesn't it check every history step? Some bell is ringing here.

Also it might be due to database lookups for some history list candy stuff. I didn't observe the issue here btw so i  was interested if there is some "special stuff" involved. Do you have an example xmp?

**Comment by apostel338** [comment]
XMP is in OP's google drive link. 

I'm so free to insert it here, if you don't want to download the zip with DNG+XMP.

[DSC00677.dng.xmp.txt](https://github.com/darktable-org/darktable/files/12643457/DSC00677.dng.xmp.txt)

**Comment by jenshannoschwalm** [comment]
Sorry - i oversaw you mentioned the image file plus xmp.

Mmh, i checked the file and also tested compressing history...

1. i couldn't see any performance / "lagginess" difference at all
2. For me the system didn't feel different
3. i checked my logs and couldn't see anything "suspicious"

Just some remarks. You are using a linear sraw dng file (probably converted ...) These are somewhat slower in some parts of the pixelpipe (opposed hlr for example) also slightly less efficient with pixelpipe cache ...

But - overall - i simply couldn't reproduce here. I would be very interested in a log, please use just '-d common' that will be sufficient.

**Comment by hafzaal** [comment]
> Sorry - i oversaw you mentioned the image file plus xmp.
> 
> Mmh, i checked the file and also tested compressing history...
> 
>     1. i couldn't see any performance / "lagginess" difference at all
> 
>     2. For me the system didn't feel different
> 
>     3. i checked my logs and couldn't see anything "suspicious"
> 
> 
> Just some remarks. You are using a linear sraw dng file (probably converted ...) These are somewhat slower in some parts of the pixelpipe (opposed hlr for example) also slightly less efficient with pixelpipe cache ...
> 
> But - overall - i simply couldn't reproduce here. I would be very interested in a log, please use just '-d common' that will be sufficient.

Thank you for looking into this. Let me know if this is sufficient or if you need any other information.

[darktable-log.txt](https://github.com/darktable-org/darktable/files/12655201/darktable-log.txt)

**Comment by jenshannoschwalm** [comment]
Yes I now see what you mean and can confirm. From my observation I would say the issue is "stuttering gtk  slider updating with large histories"

**Comment by jenshannoschwalm** [comment]
BTW this can also lead to issue described in #16160
